cmd/pipeline/main.go

```diff
@@ -632,7 +632,6 @@ func main() {
 	base.GET("version", gin.WrapH(buildinfo.Handler(buildInfo)))
```

Comment:
Starting a thread from [this message](https://github.com/banzaicloud/pipeline/pull/3419#issuecomment-778573276)
> I don't mind this approach since in this case there is actually no chance of a massive history buildup. I'm not sure this will work in a scenario where you need to iterate through the complete list of clusters ever time though.

I can derive two immediate concerns from your message and example cluster list iteration case.
I'm going to attempt to address those, but if you see other possible issues with this approach, let me know as I might have misunderstood your original intent.

1. Issue of building up too large state information during one workflow run because of a lengthy cluster list and/or many activities in a cron workflow.
2. Issue of recording current state in a (partitioned) single cron iteration logic.

I see two methods tackling these:
#### A, initiating child workflows from the cron workflow to break down the cluster list size and/or activity number. 

As parent and child history are [separated](https://cadenceworkflow.io/docs/concepts/workflows/#child-workflow), this can prevent the issue of too large history in the cron workflow regardless of the cluster/activity count.
The main pros of this approach are 
- the intuitive, easy to follow logical hierarchy of the workflows/activities,
- the simple implementation (child workflows can have a parametrized generic implementation, 
- the support of "out-of-the-box" parallelism through child workflow futures, and 
- the (cron) workflow requires few-to-none cron-run-specific logic (the degree depending on the state recording solution).

In this solution the necessary input state can be carried from the cron workflow through the input arguments of the child workflows. The current (completion) state of the single cron iteration can be determined by the state of the child workflows (which could be challenging without aggregation so) for observability purposes we can introduce additional recording measures in the parent cron workflow.

#### B, splitting up the single cron workflow run into multiple executions (runs) by returning `ContinueAsNewError` from the single cron workflow run at the partitioning points of the single iteration logic in order to reduce the single iteration history size. 

This approach is more fit for tackling the too many clusters in the list case, but would be possible - although not recommended - for the too many activities case. At certain points of the single iteration logic - let's say every 100 cluster - we could "reset" the history using manual workflow "reset", if we could carry over the state information.
The main pro of this approach is the reduced hierarchy and fewer workflows/activities, thus easier "overview" of all running workflows/activities in a system (one cron iteration is encapsulated in its single corresponding workflow).
IMO aside from the custom logic maintenance cost another drawback is the cumbersome nature of handling the "too many activities" case (the logical jumps become hard to follow for readers) and also the loss of parallelism (as only one instance of the cron workflow can be executed at any moment).

This method requires additional state recording measures in the cron workflow.
One solution is to use the workflow input arguments which can be modified through the `ContinueAsNewError` returned from the cron workflow at the partitioning points.

---

I personally would prefer method A, because it relies more on the higher level Cadence mechanics and relegates the lower level implementation to Cadence, supports parallelism and avoids additional custom workaround logic. I think solution A is intuitive, easier to modify or extend it in the future if needed while also have much better separation of responsibilities compared to solution B which also eases maintenance in the long run.

Reply:
I agree, method A would be better, but it doesn't apply here, so there is nothing else to do in my opinion.